src/main.rs

```diff
@@ -4,70 +4,58 @@ use utf8parse::Receiver;
 use std::env;
 use std::fs::File;
 use std::io::Read;
+use std::str;
 use std::time::{Duration, Instant};
 
-struct BufferReceiver {
-    data: Vec<char>,
-}
-
-impl BufferReceiver {
-    pub fn new() -> Self {
-        BufferReceiver { data: vec![] }
-    }
-    pub fn done(self) -> Vec<char> {
-        self.data
-    }
-}
+struct Utf8Receiver;
 
-impl Receiver for BufferReceiver {
-    fn codepoint(&mut self, c: char) {
-        self.data.push(c);
+impl Receiver for Utf8Receiver {
+    fn codepoint(&mut self, _c: char) {
     }
     fn invalid_sequence(&mut self) {
         panic!("This is a performance benchmark. Use only valid UTF-8");
     }
 }
 
 trait BenchableUtf8Parser {
-    fn parse(bytes: &[u8]) -> Vec<char>;
+    fn parse(bytes: Vec<u8>) -> String;
 }
 
 struct BenchableTableParser;
 
 impl BenchableUtf8Parser for BenchableTableParser {
-    fn parse(bytes: &[u8]) -> Vec<char> {
+    fn parse(bytes: Vec<u8>) -> String {
         let mut p = utf8parse::Parser::new();
-        let mut r = BufferReceiver::new();
+        let mut r = Utf8Receiver;
         for b in bytes.iter() {
             p.advance(&mut r, *b);
         }
-        r.done()
+
+        // if we didn't panic, then utf8 is valid
+        unsafe {
+            String::from_utf8_unchecked(bytes)
+        }
     }
 }
 
 struct BenchableStdlibParser;
 
 impl BenchableUtf8Parser for BenchableStdlibParser {
-    fn parse(bytes: &[u8]) -> Vec<char> {
-        let mut ret = Vec::new();
```

Comment:
I agree we should try to remove allocation overhead from the tests. As a first approximation I tried to have the same allocation overhead for both tests, using an explicit `Vec::push` instead of just a `collect()` which may do a more clever allocation using size hints.
To actually get rid of the allocation we could either just count the characters in both cases, or we could allocate the vector `with_capacity` of the input file size because there cannot be more characters than bytes in an UTF-8 encoded file. This is quite memory intensive, as a `char` is four bytes large.